frontend/app/src/components/defi/Lending.vue

```diff
@@ -126,20 +198,77 @@ export default class Lending extends Vue {
     );
   }
 
-  readonly decimals: number = 8;
+  get totalDeposited(): BigNumber {
+    return this.dsrBalances.reduce(
+      (sum, { balance }) => sum.plus(balance.amount),
+      Zero
+    );
+  }
 
-  get totalDai(): BigNumber {
+  get totalDepositedUsd(): BigNumber {
     return this.dsrBalances.reduce(
-      (sum, { balance }) => sum.plus(balance),
+      (sum, { balance }) => sum.plus(balance.usdValue),
       Zero
     );
   }
 
-  async mounted() {
-    await this.$store.dispatch('balances/fetchDSRBalances');
-    if (this.premium) {
-      await this.$store.dispatch('balances/fetchDSRHistory');
+  get totalGainAcrossFilteredAccounts(): BigNumber {
+    const filteredAccounts = this.filteredAccounts.map(
+      account => account.address
+    );
+
+    const accountsMap = Object.keys(this.dsrHistoryByAddress).map(account => {
+      return {
+        address: account,
+        ...this.dsrHistoryByAddress[account]
+      };
+    });
+
+    // Map over all of our dsrHistory, passing those that match the
+    // filtered accounts to a reduce to aggregate. If no accounts are
+    // filtered then aggregate all of them.
+    if (filteredAccounts.length !== 0) {
+      return accountsMap
+        .filter(account => filteredAccounts.indexOf(account.address) > -1)
+        .reduce((sum, account) => {
+          return sum.plus(account.gainSoFar);
+        }, Zero);
+    }
+    return accountsMap.reduce((sum, account) => {
+      return sum.plus(account.gainSoFar);
+    }, Zero);
+  }
+
+  get totalUsdGainAcrossFilteredAccounts(): BigNumber {
+    const filteredAccounts = this.filteredAccounts.map(
+      account => account.address
+    );
+
+    const accountsMap = Object.keys(this.dsrHistoryByAddress).map(account => {
+      return {
+        address: account,
+        ...this.dsrHistoryByAddress[account]
+      };
+    });
+
+    // Map over all of our dsrHistory, passing those that match the
+    // filtered accounts to a reduce to aggregate. If no accounts are
+    // filtered then aggregate all of them.
+    if (filteredAccounts.length !== 0) {
+      return accountsMap
+        .filter(account => filteredAccounts.indexOf(account.address) > -1)
+        .reduce((sum, account) => {
+          return sum.plus(account.gainSoFarUsdValue);
+        }, Zero);
     }
+    return accountsMap.reduce((sum, account) => {
+      return sum.plus(account.gainSoFarUsdValue);
```

Comment:
Can we somehow extract a common base for this and the previous function? the only difference is the property used but they are 90% the same

Reply:
I tried to do this and typescript gave me a lot of problems (I couldn't figure out how to satisfy typescript's requirement that the property being aggregated is "included" in the object). If you know how to do it I would be glad to fix it!